rollout/trafficrouting/openshift/openshift.go

```diff
@@ -0,0 +1,177 @@
+	errChan := make(chan error)
+	defer close(errChan)
+	var wg sync.WaitGroup
+	wg.Add(len(routeNameList))
+	for _, routeName := range routeNameList {
+		go func(routeName string) {
+			defer wg.Done()
+			err := r.updateRoute(ctx, routeName, desiredWeight)
+			if err != nil {
+				errChan <- err
+			}
+		}(routeName)
+	}
+	go func() {
+		wg.Wait()
+		close(doneChan)
+	}()
+
+	errs := []error{}
+	done := false
+	for !done {
+		select {
+		case <-doneChan:
+			done = true
+		case err := <-errChan:
+			errs = append(errs, err)
+		}
+	}
+	return formatErrors(errs)
+}
+
+func formatErrors(errs []error) error {
+	errorsCount := len(errs)
+	if errorsCount == 0 {
+		return nil
+	} else if errorsCount == 1 {
+		return errs[0]
+	}
+	var errMsg strings.Builder
+	errMsg.WriteString(fmt.Sprintf("%d errors found: ", errorsCount))
+	for _, err := range errs {
+		errMsg.WriteString(fmt.Sprintf("[%s]", err))
+	}
+	return errors.New(errMsg.String())
+}
+
+func (r *Reconciler) getRoute(ctx context.Context, routeName string) (*routev1.Route, error) {
+	return r.Cfg.Client.RouteV1().Routes(r.Cfg.Rollout.GetNamespace()).Get(ctx, routeName, metav1.GetOptions{})
+}
+
+//Update default backend weight,
+//remove alternateBackends if weight is 0,
+//otherwise update alternateBackends
+func (r *Reconciler) updateRoute(ctx context.Context, routeName string, desiredWeight int32) error {
+	// get the route
+	route, err := r.getRoute(ctx, routeName)
+	if err != nil {
+		if k8serrors.IsNotFound(err) {
+			msg := fmt.Sprintf("Route %q not found", routeName)
+			r.sendWarningEvent("OpenshiftRouteNotFound", msg)
+		}
+		return err
+	}
+
+	// update default backend weight
+	altWeight := 100 - desiredWeight
+	r.Log.Infof("updating default backend weight to %d", altWeight)
+	route.Spec.To.Weight = &altWeight
+	if desiredWeight == 0 {
+		r.Log.Infof("deleting alternateBackends")
+		route.Spec.AlternateBackends = nil
+	} else {
+		r.Log.Infof("updating alternate backend weight to %d", desiredWeight)
+		route.Spec.AlternateBackends = []routev1.RouteTargetReference{{
+			Kind:   "Service",
+			Name:   r.Cfg.Rollout.Spec.Strategy.Canary.CanaryService,
+			Weight: &desiredWeight,
+		}}
+	}
+
+	_, err = r.Cfg.Client.RouteV1().Routes(r.Cfg.Rollout.GetNamespace()).Update(ctx, route, metav1.UpdateOptions{})
+
+	return err
+}
+
+// Verifies weight of routes given by rollout
+func (r *Reconciler) VerifyWeight(desiredWeight int32) (bool, error) {
+	r.sendNormalEvent("VerifyWeight", fmt.Sprintf("Verify weight is %d", desiredWeight))
+	for _, routeName := range r.Cfg.Rollout.Spec.Strategy.Canary.TrafficRouting.Openshift.Routes {
+		route, err := r.getRoute(context.TODO(), routeName)
+		if err != nil {
+			return false, err
+		}
+		if route.Spec.AlternateBackends == nil {
+			if desiredWeight != 0 {
+				return false, nil
+			}
+		} else if *route.Spec.To.Weight != 100-desiredWeight || *route.Spec.AlternateBackends[0].Weight != desiredWeight {
+			return false, nil
+		}
+	}
+	return true, nil
+}
```

Comment:
You can remove this implementation and make VerifyWeight a no-op. The only reason to have a VerifyWeight, is if we need a mechanism to talk to some **_external_** system to verify the canary weights took place. Since OpenShift routes are all done against K8s APIs, we don't have some external system to verify.

The only traffic implementation which needs this right now is ALB canary, which needs to make AWS API calls to verify the LoadBalancer target groups are updated (since there is a delay between the ingress modification vs. the corresponding ALB change).

I would delete this code because the automatic verification/remediation of the Route, can be performed in SetWeight code. We don't need a separate function to verify the change that we just made.